### PR TITLE
File IO: File open/save dialog utility functions

### PR DESCRIPTION
**Pull request by Scepheo**
PR to create and use utility functions for the file open/save dialog, in a more "C++ way" than the current code.

### CODE REVIEW

File: source/application/AVIDumping/AVIDumper.cpp [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2015- Hourglass Resurrection Team
+ * Copyright (c) 2017 - Hourglass Resurrection Team
```
Comment (YaLTeR) [inline source/application/AVIDumping/AVIDumper.cpp]: Pretty sure this should be 2015-2017.

File: source/application/utils/File.h [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+            AllFiles,
+            HourglassMovie,
+            Executable,
+            Config,
+            WatchList,
+            AVI
+        };
+
+        std::string GetFileNameOpen(
+            const std::string location,
```
Comment (YaLTeR) [inline source/application/utils/File.h]: You might've meant to pass all these by const reference, in which case you forgot the &.

File: source/application/utils/file/Path.cpp [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+std::string Utils::File::GetFileName(const std::string path)
+{
+    size_t last_slash = path.find_last_of("/\\");
+    return last_slash == std::string::npos ? path : path.substr(last_slash + 1);
+}
+
+std::string Utils::File::GetDirectoryName(const std::string path)
+{
+    size_t last_slash = path.find_last_of("/\\");
+    return last_slash == std::string::npos ? std::string() : path.substr(0, last_slash);
```
Comment (YaLTeR) [inline source/application/utils/file/Path.cpp]: This won't work the same on `C:\Folder\` vs. `C:\Folder`, is that the intended behaviour?

File: source/application/wintaser.cpp [hunk 4]
```diff
@@ -1746,34 +1747,32 @@ void LoadGameStatePhase2(int slot)
...
-	// If hGameProcess has a value at this point and file creation was successful,
-	// register hGameProcess with the AVI Dumper now so that we can start capture immediately
-	if(hGameProcess) SetCaptureProcess(hGameProcess);
-	return true;
+    std::string avi_file_name = Utils::File::GetFileNameSave(thisprocessPath, {
+        Utils::File::FileFilter::AVI,
+        Utils::File::FileFilter::AllFiles
+    });
+
+    if (avi_file_name == "")
```
Comment (YaLTeR) [inline source/application/wintaser.cpp]: I think `avi_file_name.empty()` would  be clearer here.

File: source/application/AVIDumping/AVIDumper.cpp (hunk 1 above)
Comment (Warepire) [reply source/application/AVIDumping/AVIDumper.cpp]: That should have remained simply "2015-".

File: source/application/AVIDumping/AVIDumper.h [hunk 5]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2015- Hourglass Resurrection Team
+ * Copyright (c) 2017 - Hourglass Resurrection Team
```
Comment (Warepire) [inline source/application/AVIDumping/AVIDumper.h]: Incorrect, copyrights don't update like this. Should remain at "2015-".

File: source/application/Config.cpp [hunk 6]
```diff
@@ -1,3 +1,9 @@
+/*
+ * Copyright(c) 2017 - Hourglass Resurrection Team
```
Comment (Warepire) [inline source/application/Config.cpp]: Should probably say "2015-", since it's an older file.

File: source/application/Config.h [hunk 7]
```diff
@@ -1,3 +1,9 @@
+/*
+ * Copyright(c) 2017 - Hourglass Resurrection Team
```
Comment (Warepire) [inline source/application/Config.h]: Same here, 2015.

File: source/application/utils/file/FileDialog.cpp [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+
+#include <map>
+#include <algorithm>
+
+#include "../File.h"
+
+namespace
+{
+    // The pipe characters are just placeholders for the null characters to ease string manipulation
+    std::map<Utils::File::FileFilter, const char*> gs_file_type_map =
```
Comment (Warepire) [inline source/application/utils/file/FileDialog.cpp]: Not static, nor global.

File: source/application/utils/file/FileDialog.cpp [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+        { Utils::File::FileFilter::Config, "Config Files (*.cfg)|*.cfg|" },
+        { Utils::File::FileFilter::WatchList, "Watchlist (*.wch)|*.wch|" },
+        { Utils::File::FileFilter::AVI, "AVI file (*.avi)|*.avi|" }
+    };
+    
+    std::string GetFilter(const std::vector<Utils::File::FileFilter> file_types)
+    {
+        std::string result = "";
+
+        for (auto file_type : file_types)
```
Comment (Warepire) [inline source/application/utils/file/FileDialog.cpp]: `auto&`, ensures you're using references and not copies.

File: source/application/utils/file/Path.cpp [hunk 10]
```diff
@@ -0,0 +1,19 @@
+/*
+ * Copyright(c) 2017 - Hourglass Resurrection Team
+ * Hourglass Resurrection is licensed under GPL v2.
+ * Refer to the file COPYING.txt in the project root.
+ */
+
+#include "../File.h"
+
+std::string Utils::File::GetFileName(const std::string path)
```
Comment (Warepire) [inline source/application/utils/file/Path.cpp]: I've been thinking about this, what about a class called `Path` that inherits `std::string`?

File: source/application/utils/file/Path.cpp (hunk 3 above)
Comment (Warepire) [reply source/application/utils/file/Path.cpp]: @YaLTeR I believe this was broken also in the original code.

File: source/application/wintaser.cpp (hunk 4 above)
Comment (Warepire) [reply source/application/wintaser.cpp]: As YaLTeR said, test against `.empty()`

File: source/application/wintaser.cpp [hunk 11]
```diff
@@ -6448,11 +6439,7 @@ BOOL CALLBACK DlgProc(HWND hDlg, UINT message, WPARAM wParam, LPARAM lParam)
 					{
 				//case ID_FILES_RESUMEMOVAS: // TODO: Eliminate this case
 						Save_Config();
-						char filename [MAX_PATH+1] = {0};
-						char directory [MAX_PATH+1] = {0};
-						SplitToValidPath(moviefilename, thisprocessPath, filename, directory);
-						NormalizePath(moviefilename, moviefilename);
-
+
```
Comment (Warepire) [inline source/application/wintaser.cpp]: Lot's of whitespace for no reason.

**Review by Warepire — CHANGES_REQUESTED** [review]

File: source/application/utils/file/Path.cpp (hunk 3 above)
Comment (Scepheo) [reply source/application/utils/file/Path.cpp]: @Warepire not broken, intended behaviour. `C:\Folder\` is the folder called `Folder` under `C:`, `C:\Folder` is the file called `Folder`. Same goes for `GetFileName`; this is simply how Windows deals with paths.

File: source/application/utils/file/Path.cpp (hunk 10 above)
Comment (Scepheo) [reply source/application/utils/file/Path.cpp]: Not entirely sure what you mean, do you just want extra type safety?

File: source/application/utils/file/Path.cpp (hunk 3 above)
Comment (Warepire) [reply source/application/utils/file/Path.cpp]: Right. We might need some sugar on top for any prompt where we should pick a path later.

File: source/application/utils/file/Path.cpp (hunk 10 above)
Comment (Warepire) [reply source/application/utils/file/Path.cpp]: I mean, `class Path : public std::string` that gives all the functionality of string (and can be passed as a string if we don't add member variables), but with these additional functions to get directory and file name from the class.

File: .gitignore [hunk 12]
```diff
@@ -9,3 +9,4 @@ build/
 *.VC.opendb
 *.VC.db
 ipch/
+.vs/
```
Comment (Warepire) [inline .gitignore:12]: This could preferably be it's own commit.

File: source/application/utils/file/Path.cpp [hunk 13]
```diff
@@ -0,0 +1,19 @@
+/*
+ * Copyright(c) 2017- Hourglass Resurrection Team
+ * Hourglass Resurrection is licensed under GPL v2.
+ * Refer to the file COPYING.txt in the project root.
+ */
+
+#include "../File.h"
+
+std::string Utils::File::GetFileName(const std::string& path)
```
Comment (Warepire) [inline source/application/utils/file/Path.cpp]: I'm still interested in discussing whether the Path utils should be a class inheriting from `std::string` and these being additional member functions.

**Review by Warepire — CHANGES_REQUESTED** [review]

**Comment by Scepheo** [comment]
Due to the move to C++17 and VS2017, I'll be dropping this PR and seeing if the changes still apply. If they do, I'll open a new PR to merge them.

**Pull request closed without merging**